packages/field-plugin/src/createFieldPlugin/createPluginActions/callbackQueue.ts

```diff
@@ -0,0 +1,66 @@
+import {
+  AssetSelectedMessage,
+  ContextRequestMessage,
+  OnMessage,
+  LoadedMessage,
+} from '../../messaging'
+
+type CallbackId = string
+type CallbackRef<Message> = {
+  callbackId: CallbackId
+  callback: OnMessage<Message>
+}
+type CallbackMap = {
+  asset: CallbackRef<AssetSelectedMessage>[]
+  context: CallbackRef<ContextRequestMessage>[]
+  state: CallbackRef<LoadedMessage>[]
```

Comment:
Let's say

```js
const plugin1 = createFieldPlugin();
const plugin2 = createFieldPlugin();
const plugin3 = createFieldPlugin();

...

plugin1.actions.setContent({...});
```

After that, the container will send the new state. In this case, I think all of plugin 1 ~ 3 should receive the state update regardless of callbackId. What do you think?

Reply:
Yes, all of them would. And the promise that was created in one of them will be resolved. These lines are for handling the callback function in the promise.

Reply:
Oh, I see. All the plugins get updated by `onLoaded` event, but only the caller's promise gets resolved by the callback.

I thought valueChangeMessage and modalChangeMessage contained the states, but they're just return value from the Container, and I assume the Container send a separate `loaded` event, so that all the plugins' `onLoaded` event gets triggered?

Reply:
`onLoaded` is called only once after the library is loaded.

And there's `onStateChange`, which can be returned after `setContent` or `setModalOpen`.

https://github.com/storyblok/field-plugin/pull/260/files#diff-e62bdc8d2bfaac8dcd9c0ec5f420b868131bc32f10c4be5743ba3e5e28ebe938R49-R54

If `plugin1` calls `setContent`, and Storyfront will `stateChanged` event back to all of plugin1~3. Only plugin1 will resolve the promise. However, all of plugin1~3 will still call the updater function.

```js
    popCallback('stateChanged', data.callbackId)?.(data)
    onUpdateState(pluginStateFromStateChangeMessage(data))
```

I was writing down a question, and in the middle of it, I understood. But still left it as a record for others :)